compiler-rt/lib/scudo/standalone/secondary.h

```diff
@@ -334,61 +336,110 @@ class MapAllocatorCache {
         const uptr HeaderPos = AllocPos - HeadersSize;
-        if (HeaderPos > CommitBase + CommitSize)
-          continue;
-        if (HeaderPos < CommitBase ||
-            AllocPos > CommitBase + PageSize * MaxUnusedCachePages) {
+        if (HeaderPos > CommitBase + CommitSize || HeaderPos < CommitBase)
           continue;
-        }
-        Found = true;
+
         const uptr Diff = HeaderPos - CommitBase;
-        // immediately use a cached block if it's size is close enough to the
-        // requested size.
-        const uptr MaxAllowedFragmentedBytes =
-            (CommitBase + CommitSize - HeaderPos) / FragmentedBytesDivisor;
-        if (Diff <= MaxAllowedFragmentedBytes) {
-          OptimalFitIndex = I;
-          EntryHeaderPos = HeaderPos;
-          break;
-        }
-        // keep track of the smallest cached block
-        // that is greater than (AllocSize + HeaderSize)
-        if (Diff > MinDiff)
+
+        if (Diff > MaxAllowedFragmentedBytes || Diff >= MinDiff)
           continue;
-        OptimalFitIndex = I;
+
         MinDiff = Diff;
+        OptimalFitIndex = I;
         EntryHeaderPos = HeaderPos;
+
+        const uptr OptimalFitThesholdBytes =
+            (CommitBase + CommitSize - HeaderPos) / FragmentedBytesDivisor;
+        if (Diff <= OptimalFitThesholdBytes) {
+          FoundOptimalFit = true;
+          break;
+        }
       }
-      if (Found) {
+      if (OptimalFitIndex != CachedBlock::InvalidEntry) {
         Entry = Entries[OptimalFitIndex];
         remove(OptimalFitIndex);
         SuccessfulRetrieves++;
       }
     }
 
+    //  The difference between the retrieved memory chunk and the request
+    //  size is at most MaxAllowedFragmentedBytes
+    //
+    //  /     MaxAllowedFragmentedBytes      \
+    // +--------------------------+-----------+
+    // |                          |           |
+    // +--------------------------+-----------+
+    //  \ Bytes to be released   /      ^
+    //                                  |
+    //                         (may or may not have commited)
+    //
+    //   The maximum number of bytes released to the OS is capped by
+    //   ReleaseMemoryUpperBound
+    //
+    //   * ReleaseMemoryUpperBound default is currently 16 KB
+    //        - We arrived at this value after noticing that mapping
+    //        in larger memory regions performs better than releasing
+    //        memory and forcing a cache hit. According to the data,
+    //        it suggests that beyond 16KB, the release execution time is
+    //        longer than the map execution time. In this way, the default
+    //        is dependent on the platform.
+    //
+    //   TODO : Considering to make ReleaseMemoryUpperBound configurable since
+    //   the release to OS API can vary across systems.
+    if (!FoundOptimalFit && Entry.Time != 0) {
+      const uptr FragmentedBytes = EntryHeaderPos - Entry.CommitBase;
```

Comment:
About `EntryHeaderPos`, do we need to round down to the page boundary?

Reply:
In the current implementation, I don't think we need to because we guarantee that there are at least `MaxUnusedCacheBytes` (which is `4 * PageSize`) between the memory being released and `EntryHeaderPos` (this is done with the check `if (FragmentedBytes > MaxUnusedCacheBytes)`). However, in the future if `MaxUnusedCacheBytes < PageSize` or we decide to release memory closer to the header, this would be necessary. Do you think it's worth adding now?

Reply:
Suppose the commitBase is at addr 0, EntryHeaderPos is at 4 * PageSize + 4080 (the header sits on the fifth page from the base). So the BytesToRelease will be 4096 (the page size). But in this case, it can be 0, right?

BTW, MaxUnusedCacheBytes shouldn't be less than a page (nor not a multiple of page size) because the mapping/releasing align at the page boundary.

Reply:
Yes, `BytesToRelease` would be 4096 in that case. I assumed that if there was any excess memory beyond `MaxUnusedCacheBytes`, then we should have a release, but since the release API releases pages, we round up to the nearest page. In any case, it still shouldn't touch `EntryHeaderPos` given that there is always at least one page between `EntryHeaderPos` and the base of any released page. Is there another concern that I may be missing?

You're right, `MaxUnusedCacheBytes` will always be a multiple of   `PageSize` right now. I was just saying that in the future if we set the unused bytes to something that wasn't a multiple of page size (highly unlikely), this is a case where we may accidentally release the header page.

Reply:
The problem here is that we may release one additional page that are not supposed to be released. In the case I mentioned, we can avoid doing any page release but we end up having one instead. When the `MaxUnusedCacheBytes` is 0, then we touch the page where the header is (Of course it's fine because the header is not set yet)

Reply:
If the header is at 4 * PageSize + 4080, though, I intentionally release one page because we still technically exceed `MaxUnusedCacheBytes = MaxUnusedCachePages * PageSize = 4 * PageSize` unused bytes. Technically, there aren't 5 unused pages because the 5th page is being used by the header, so I don't see a problem with rounding down `EntryHeaderPos`, but I just wanted to point out why I was initially doing a release in this case. Either way sounds good to me, though.

Reply:
The fifth page will be used by the header (it's not unused), the unused pages are 4. It doesn't exceed the `MaxUnusedCacheBytes`. The issue is how we calculate the *unused pages* and that's why I think we should do a round down.

Reply:
Ok, sounds good to me.

Reply:
@ChiaHungDuan , just wanted to highlight that the computation of `Diff` was changed to accomodate this change.